File: sdk/runanywhere-web/packages/core/src/Public/API/Results.ts
/**
 * Result types of the v4 public API.
 *
 * Nothing here carries a `success` flag or an error message in a text field —
 * failures throw a typed [SDKException].
 */

import type { ModelCategory, ModelInfo } from '@runanywhere/proto-ts/model_types';
import type { ToolCall } from '@runanywhere/proto-ts/tool_calling';
import type { ProtoSDKError } from '../../Foundation/SDKException.js';
import type { AudioFrame } from './Inputs.js';
import type { Backend, BackendPreference, StructuredOutputMode } from './Options.js';
import type { TranscriptionEvent, VadEvent } from './Events.js';

/** Why generation stopped. */
export type FinishReason = 'stop' | 'length' | 'toolCalls' | 'cancelled' | 'contentFilter' | 'unknown';

/** Throughput and token accounting shared by every generation result. */
export interface GenerationMetrics {
  inputTokens: number;
  outputTokens: number;
  timeToFirstTokenMs: number;
  tokensPerSecond: number;
  requestId: string;
  model: string;
}

/** Completed text generation. */
export interface GenerationResult extends GenerationMetrics {
  text: string;
  thinkingText?: string;
  toolCalls: ToolCall[];
  finishReason: FinishReason;
  /** Backend-native finish-reason string, before normalization into [FinishReason]. */
  rawFinishReason?: string;
  actualBackend?: Backend;
  actualDevice?: string;
}

/** Completed structured generation, with the parsed value alongside the raw text. */
export interface StructuredResult extends GenerationMetrics {
  value: unknown;
  raw: string;
  valid: boolean;
  mode: StructuredOutputMode;
}

/** One recognized word of a transcription. */
export interface Word {
  text: string;
  startMs: number;
  endMs: number;
  confidence: number;
  speakerId?: string;
}

/** Completed transcription. */
export interface Transcription {
  text: string;
  language?: string;
  confidence: number;
  words: Word[];
  durationMs: number;
}

/** Synthesized audio. */
export interface Audio {
  data: Uint8Array;
  sampleRate: number;
  format: 'pcm' | 'pcm16' | 'wav';
  durationMs: number;
}

/** One chunk of a streamed synthesis. */
export interface AudioChunk {
  data: Uint8Array;
  index: number;
  isFinal: boolean;
}

/** A span of detected speech. */
export interface Segment {
  startMs: number;
  endMs: number;
}

/** Speech-detection verdict for one audio payload. */
export interface VadResult {
  isSpeech: boolean;
  probability: number;
  segments: Segment[];
}

/** One embedding, tagged with the index of its input text. */
export interface Embedding {
  index: number;
  vector: Float32Array;
}

/** One reranked document, pointing back at its input index. */
export interface RankedResult {
  index: number;
  relevanceScore: number;
}

/** One generated image. */
export interface ImageData {
  bytes: Uint8Array;
  width: number;
  height: number;
}

/** Completed image generation. */
export interface ImageResult {
  images: ImageData[];
  seed: number;
  steps: number;
}

/** One diarized span attributed to a speaker. */
export interface SpeakerSegment {
  speakerId: string;
  startMs: number;
  endMs: number;
}

/** Completed speaker diarization. */
export interface DiarizationResult {
  segments: SpeakerSegment[];
  speakerCount: number;
}

/** One class present in a segmentation mask. */
export interface ClassInfo {
  id: number;
  label: string;
  pixelCount: number;
  fraction: number;
}

/** Completed semantic segmentation. */
export interface SegmentationResult {
  /** Per-pixel class ids, u16 little-endian, row-major. */
  classMask: Uint8Array;
  width: number;
  height: number;
  classes: ClassInfo[];
  /** RGBA preview, present when `includeDiagnosticImage` was set. */
  diagnosticImage?: Uint8Array;
}

/** One retrieved chunk of a RAG corpus. */
export interface Match {
  text: string;
  score: number;
  metadata: Record<string, string>;
}

/** Completed RAG answer with the context it was grounded in. */
export interface RagResult extends GenerationMetrics {
  answer: string;
  sources: Match[];
}

/** Readiness of the loaded speech-to-text model. */
export interface SttState {
  isReady: boolean;
  modelId?: string;
  supportsStreaming: boolean;
  languages: string[];
}

/** One voice the loaded synthesis model can render. */
export interface Voice {
  id: string;
  name: string;
  language: string;
  isNeural: boolean;
}

/** What is resident and how much room is left. */
export interface ModelsState {
  loaded: Partial<Record<ModelCategory, ModelInfo>>;
  storageUsedBytes: number;
  storageFreeBytes: number;
}

/** One applied LoRA adapter. */
export interface AppliedAdapter {
  id: string;
  scale: number;
}

/** Adapters currently layered onto the base model. */
export interface LoraState {
  applied: AppliedAdapter[];
}

/** Size of a RAG corpus. */
export interface RagStats {
  documentCount: number;
  chunkCount: number;
  indexSizeBytes: number;
}

/**
 * Ownership handle for one resident model, returned by `models.load`.
 *
 * `close()` and `models.unload(id)` release the same residency; calling
 * either after the other is a no-op.
 */
export interface LoadedModel {
  readonly id: string;
  readonly category: ModelCategory;
  /** The backend preference that was requested, when one was given. */
  readonly requestedBackend?: BackendPreference;
  readonly actualBackend: Backend;
  readonly actualDevice: string;
  readonly runtimeVersion?: string;
  readonly abiVersion?: string;
  /** Set when the requested backend/accelerator could not be honored. */
  readonly fallbackReason?: string;
  /** Release this model's residency. Idempotent. */
  close(): Promise<void>;
}

/**
 * Handle to one in-flight or completed `tts.speak`/`VoiceSession.say` utterance.
 */
export interface SpeechHandle {
  readonly id: string;
  readonly interrupted: boolean;
  readonly error?: ProtoSDKError;
  /** Stop playback and any in-flight synthesis. Resolves once stopped. */
  interrupt(): Promise<void>;
  /** Resolve once playback finishes, is interrupted, or fails. */
  waitForPlayout(): Promise<void>;
}

/**
 * Live speech-to-text session opened by `stt.openStream`. Establishes its
 * audio format once; every pushed frame carries PCM samples in that format.
 */
export interface SttStream {
  readonly events: AsyncIterable<TranscriptionEvent>;
  /** Push one frame of PCM audio in the stream's established format. */
  pushFrame(frame: AudioFrame): void;
  /** Request the backend surface partials for audio pushed so far. */
  flush(): void;
  /** Signal that no more audio is coming; the backend finalizes the transcript. */
  finish(): void;
  /** Release the stream's resources. Idempotent. */
  close(): Promise<void>;
}

/**
 * Live voice-activity session opened by `vad.openStream`. Establishes its
 * audio format once; every pushed frame carries PCM samples in that format.
 */
export interface VadStream {
  readonly events: AsyncIterable<VadEvent>;
  /** Push one frame of PCM audio in the stream's established format. */
  pushFrame(frame: AudioFrame): void;
  /** No-op on Web: there is no partial-result buffer to flush. */
  flush(): void;
  /** Signal that no more audio is coming; completes the event stream. */
  finish(): void;
  /** Release the stream's resources. Idempotent. */
  close(): Promise<void>;
}

/** Whether a modality/backend/feature is honestly available right now. */
export interface UnavailableCapability {
  name: string;
  reason: string;
}

/** Per-modality streaming support, keyed by namespace name. */
export interface StreamingCapabilities {
  llm: boolean;
  vlm: boolean;
  stt: boolean;
  tts: boolean;
  vad: boolean;
  rag: boolean;
  images: boolean;
}

/** Tool-calling support of the currently registered backends. */
export interface ToolCapabilities {
  registry: boolean;
  parallel: boolean;
  cancellation: boolean;
}

/** RAG-session support of the currently registered backends. */
export interface RagCapabilities {
  multiSession: boolean;
  persistent: boolean;
}

/**
 * Installed, packaged, and executable surface of this SDK build, generated
 * from packaging and runtime probes rather than from IDL enum presence
 * alone. `capabilities()` is the source of truth apps should consult before
 * calling into a modality that might not ship on this platform.
 */
export interface SDKCapabilities {
  modalities: string[];
  backends: Backend[];
  audioFormats: string[];
  streaming: StreamingCapabilities;
  tools: ToolCapabilities;
  rag: RagCapabilities;
  unavailable: UnavailableCapability[];
}

export type { ModelInfo };

File: bindings/kotlin/src/main/kotlin/com/runanywhere/sdk/public/api/Results.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public v3 surface: every result shape, with one shared metrics block on the
 * generation results so no caller computes throughput itself.
 */

package com.runanywhere.sdk.public.api

import com.runanywhere.sdk.foundation.errors.SDKException
import kotlinx.coroutines.CompletableDeferred
import java.util.UUID
import java.util.concurrent.atomic.AtomicBoolean

/** Why a generation stopped. */
public enum class FinishReason {
    STOP,
    LENGTH,
    TOOL_CALLS,
    CANCELLED,
    CONTENT_FILTER,
    ERROR,
    UNKNOWN,
}

/** A completed text or vision generation, with its metrics. */
public data class GenerationResult(
    val text: String,
    val thinkingText: String? = null,
    val toolCalls: List<ToolCall> = emptyList(),
    /** Outcomes of the tool calls the SDK executed during the loop. */
    val toolResults: List<ToolResult> = emptyList(),
    val finishReason: FinishReason = FinishReason.STOP,
    /** Backend-native finish-reason string, before normalization into [finishReason]. */
    val rawFinishReason: String? = null,
    val inputTokens: Int = 0,
    val outputTokens: Int = 0,
    val timeToFirstTokenMs: Long = 0L,
    val tokensPerSecond: Float = 0f,
    val requestId: String = "",
    val model: String = "",
    val actualBackend: Backend? = null,
    val actualDevice: String? = null,
)

/** A generation parsed against a JSON schema, with the same metrics as [GenerationResult]. */
public data class StructuredResult(
    /** Parsed JSON, as the canonical text commons produced. */
    val value: String,
    val raw: String,
    val valid: Boolean,
    val mode: StructuredOutputMode = StructuredOutputMode.VALIDATION_ONLY,
    val inputTokens: Int = 0,
    val outputTokens: Int = 0,
    val timeToFirstTokenMs: Long = 0L,
    val tokensPerSecond: Float = 0f,
    val requestId: String = "",
    val model: String = "",
)

/** One recognised word with its timing. */
public data class Word(
    val text: String,
    val startMs: Long,
    val endMs: Long,
    val confidence: Float,
    val speakerId: String? = null,
)

/** A completed transcription. */
public data class Transcription(
    val text: String,
    val language: String? = null,
    val confidence: Float = 0f,
    val words: List<Word> = emptyList(),
    val durationMs: Long = 0L,
)

/** Synthesized audio. */
public class Audio internal constructor(
    public val data: ByteArray,
    public val sampleRate: Int,
    public val format: AudioFormat,
    public val durationMs: Long,
)

/** One chunk of a streaming synthesis. */
public class AudioChunk internal constructor(
    public val data: ByteArray,
    public val index: Int,
    public val isFinal: Boolean,
)

/** A span of speech inside an analysed buffer. */
public data class Segment(
    val startMs: Long,
    val endMs: Long,
)

/** The verdict of a voice-activity check. */
public data class VadResult(
    val isSpeech: Boolean,
    val probability: Float,
    val segments: List<Segment> = emptyList(),
)

/** One embedding vector, tagged with its position in the input list. */
public class Embedding internal constructor(
    public val index: Int,
    public val vector: FloatArray,
)

/** One reranked document, pointing back at its input position. */
public data class RankedResult(
    val index: Int,
    val relevanceScore: Float,
)

/** One generated image. */
public class ImageData internal constructor(
    public val bytes: ByteArray,
    public val width: Int,
    public val height: Int,
    public val mediaType: String,
)

/** A completed image generation. */
public data class ImageResult(
    val images: List<ImageData>,
    val seed: Long,
    val steps: Int,
)

/** One speaker's turn inside an analysed buffer. */
public data class SpeakerSegment(
    val speakerId: String,
    val startMs: Long,
    val endMs: Long,
)

/** A completed speaker diarization. */
public data class DiarizationResult(
    val segments: List<SpeakerSegment>,
    val speakerCount: Int,
)

/** One class present in a segmentation mask. */
public data class ClassInfo(
    val classId: Int,
    val label: String,
    val pixelCount: Long,
    val fraction: Float,
)

/** A completed semantic segmentation. */
public class SegmentationResult internal constructor(
    /** Per-pixel class ids, little-endian `uint16`, row-major. */
    public val classMask: ByteArray,
    public val width: Int,
    public val height: Int,
    public val classes: List<ClassInfo>,
    /** RGBA overlay, present only when `includeDiagnosticImage` was requested. */
    public val diagnosticImage: ByteArray?,
)

/** One retrieved chunk with its similarity score. */
public data class Match(
    val text: String,
    val score: Float,
    val metadata: Map<String, String> = emptyMap(),
)

/** A completed RAG answer, with the same metrics as [GenerationResult]. */
public data class RagResult(
    val answer: String,
    val sources: List<Match>,
    /** Commons-split reasoning channel (`RAGResult.thinking_content`); null when absent. */
    val thinkingText: String? = null,
    val inputTokens: Int = 0,
    val outputTokens: Int = 0,
    val timeToFirstTokenMs: Long = 0L,
    val tokensPerSecond: Float = 0f,
    val requestId: String = "",
    val model: String = "",
)

/** Readiness of the speech-to-text component. */
public data class SttState(
    val isReady: Boolean,
    val modelId: String? = null,
    val supportsStreaming: Boolean = false,
    val languages: List<String> = emptyList(),
)

/** What is loaded and how much room is left. */
public data class ModelsState(
    val loaded: Map<ModelCategory, ModelInfo>,
    val storageUsedBytes: Long,
    val storageFreeBytes: Long,
)

/** One LoRA adapter currently applied to the loaded base model. */
public data class AppliedAdapter(
    val id: String,
    val scale: Float,
)

/** Which LoRA adapters are applied. */
public data class LoraState(
    val applied: List<AppliedAdapter>,
)

/** Size of a RAG session's index. */
public data class RagStats(
    val documentCount: Long,
    val chunkCount: Long,
    val indexSizeBytes: Long,
)

/**
 * Resident-model ownership handle returned by `models.load`.
 *
 * [close] and `models.unload(id)` release the same residency; calling
 * either after the other is a no-op.
 */
public class LoadedModel internal constructor(
    public val id: String,
    public val category: ModelCategory,
    /** The backend preference that was requested, when one was given. */
    public val requestedBackend: BackendPreference? = null,
    public val actualBackend: Backend,
    public val actualDevice: String = "unknown",
    public val runtimeVersion: String? = null,
    public val abiVersion: String? = null,
    /** Set when the requested backend/accelerator could not be honored. */
    public val fallbackReason: String? = null,
    private val closeHandler: suspend (String) -> Unit,
) {
    private val closed = AtomicBoolean(false)

    /** Release this model's residency. Idempotent. */
    public suspend fun close() {
        if (!closed.compareAndSet(false, true)) return
        closeHandler(id)
    }
}

/**
 * Handle to one in-flight or completed `tts.speak`/`VoiceSession.say` utterance.
 */
public class SpeechHandle internal constructor(
    public val id: String = UUID.randomUUID().toString(),
    private val interruptHandler: suspend () -> Unit,
) {
    private val done = CompletableDeferred<Unit>()
    private val interruptedFlag = AtomicBoolean(false)

    @Volatile
    public var error: SDKException? = null
        internal set

    public val interrupted: Boolean
        get() = interruptedFlag.get()

    /** Stop playback and any in-flight synthesis. Suspends until stopped. */
    public suspend fun interrupt() {
        interruptedFlag.set(true)
        interruptHandler()
        complete()
    }

    /** Suspend until playback finishes, is interrupted, or fails. */
    public suspend fun waitForPlayout() {
        done.await()
    }

    internal fun complete(error: SDKException? = null) {
        if (error != null) this.error = error
        done.complete(Unit)
    }
}

/** How to interpret the bytes pushed to a live [SttStream]/[VadStream]. */
public data class AudioFrame(
    val samples: ByteArray,
    val sampleCount: Int,
    val timestampMs: Long? = null,
)

/**
 * Live speech-to-text session opened by `stt.openStream`. Establishes its
 * audio format once; every pushed [AudioFrame] carries PCM samples in that format.
 */
public interface SttStream {
    public val events: kotlinx.coroutines.flow.Flow<TranscriptionEvent>

    /** Push one frame of PCM audio in the stream's established format. */
    public fun pushFrame(frame: AudioFrame)

    /** Request the backend surface partials for audio pushed so far. */
    public fun flush()

    /** Signal that no more audio is coming; the backend finalizes the transcript. */
    public fun finish()

    /** Release the stream's resources. Idempotent. */
    public suspend fun close()
}

/**
 * Live voice-activity session opened by `vad.openStream`. Establishes its
 * audio format once; every pushed [AudioFrame] carries PCM samples in that format.
 */
public interface VadStream {
    public val events: kotlinx.coroutines.flow.Flow<VadEvent>

    /** Push one frame of PCM audio in the stream's established format. */
    public fun pushFrame(frame: AudioFrame)

    /** No-op when there is no partial-result buffer to flush. */
    public fun flush()

    /** Signal that no more audio is coming; completes the event stream. */
    public fun finish()

    /** Release the stream's resources. Idempotent. */
    public suspend fun close()
}

/** Whether a modality/backend/feature is honestly available right now. */
public data class UnavailableCapability(
    val name: String,
    val reason: String,
)

/** Per-modality streaming support, keyed by namespace name. */
public data class StreamingCapabilities(
    val llm: Boolean = true,
    val vlm: Boolean = true,
    val stt: Boolean = true,
    val tts: Boolean = true,
    val vad: Boolean = true,
    val rag: Boolean = true,
    val images: Boolean = true,
)

/** Tool-calling support of the currently registered backends. */
public data class ToolCapabilities(
    val registry: Boolean = true,
    val parallel: Boolean = false,
    val cancellation: Boolean = true,
)

/** RAG-session support of the currently registered backends. */
public data class RagCapabilities(
    val multiSession: Boolean = true,
    val persistent: Boolean = true,
)

/**
 * Installed, packaged, and executable surface of this SDK build, generated
 * from packaging and runtime probes rather than from IDL enum presence
 * alone. `RunAnywhere.capabilities()` is the source of truth apps should
 * consult before calling into a modality that might not ship on this platform.
 */
public data class SDKCapabilities(
    val modalities: Set<String>,
    val backends: Set<Backend>,
    val audioFormats: Set<AudioFormat>,
    val streaming: StreamingCapabilities,
    val tools: ToolCapabilities,
    val rag: RagCapabilities,
    val unavailable: List<UnavailableCapability>,
)
